src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/net/GRPCConnectionManager.java

```diff
@@ -159,28 +155,36 @@ private ManagedChannel getChannelForHost(final String remoteHost) {
    * @param remoteHost The host to which we want to establish a channel to.
    * @return a Managed channel object.
    */
-  private synchronized ManagedChannel addOrUpdateChannelForHost(final String remoteHost) {
+  private synchronized ManagedChannel addOrUpdateChannelForHost(final InstanceDetails remoteHost) {
     final ManagedChannel channel = buildChannelForHost(remoteHost);
-    perHostChannelMap.computeIfAbsent(remoteHost, s -> new AtomicReference<>());
-    perHostChannelMap.get(remoteHost).set(channel);
+    perHostChannelMap.computeIfAbsent(remoteHost.getInstanceId(), s -> new AtomicReference<>());
+    perHostChannelMap.get(remoteHost.getInstanceId()).set(channel);
     return channel;
   }
 
-  private ManagedChannel buildChannelForHost(final String remoteHost) {
+  private ManagedChannel buildChannelForHost(final InstanceDetails remoteHost) {
     return shouldUseHttps ? buildSecureChannel(remoteHost) : buildInsecureChannel(remoteHost);
   }
 
-  private ManagedChannel buildInsecureChannel(final String remoteHost) {
-    return ManagedChannelBuilder.forAddress(remoteHost, this.port).usePlaintext().build();
+  private ManagedChannel buildInsecureChannel(final InstanceDetails remoteHost) {
+    int port = this.port != 0 ? this.port : remoteHost.getGrpcPort();
+    if (port == -1) {
+      throw new IllegalArgumentException("Invalid port for grpc: " + port);
+    }
```

Comment:
nit: factor this out into a function since it's used in multiple places

Reply:
done !